crypto/rsa/rsa_gen.c

```diff
@@ -53,19 +64,41 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
         goto err;
     }
 
+    if (primes < RSA_DEFAULT_PRIME_NUM
+        || primes > RSA_MAX_PRIME_NUM || bits <= primes) {
+        RSAerr(RSA_F_RSA_BUILTIN_KEYGEN, RSA_R_KEY_PRIME_NUM_INVALID);
+        goto err;
+    }
+
     ctx = BN_CTX_new();
     if (ctx == NULL)
         goto err;
     BN_CTX_start(ctx);
     r0 = BN_CTX_get(ctx);
     r1 = BN_CTX_get(ctx);
     r2 = BN_CTX_get(ctx);
-    r3 = BN_CTX_get(ctx);
-    if (r3 == NULL)
+    if (r2 == NULL)
         goto err;
 
-    bitsp = (bits + 1) / 2;
-    bitsq = bits - bitsp;
+    /* divide bits into 'primes' pieces evenly */
+    quo = bits / primes;
+    rmd = bits % primes;
+    for (i = 0; i < primes; i++) {
+        bitsr[i] = (i < rmd) ? quo + 1 : quo;
+        if (primes > RSA_DEFAULT_PRIME_NUM && bitsr[i] < RSA_MIN_PRIME_SIZE) {
```

Comment:
This `if` can (and should) be taken out of the loop. It's sufficient to compare `quo` for boundary condition.

Reply:
Moved outside...